src/java/org/tensorics/core/tree/walking/Trees.java

```diff
@@ -197,16 +174,16 @@ public void onEvery(Node node) {
         return newRebuildingContext.getUpdatedOrSame(rootNode);
     }
 
-    /**
+	/**
      * walks through the tree, starting from the given rootNode and collects all the nodes which implement the given
      * class.
      * 
      * @param rootNode the node from which to start the search
      * @param nodeClassToFind the class of the nodes to find
      * @return a set of all found nodes, which implement the given class
      */
-    public static <T extends Node> Set<T> findNodesOfClass(Node rootNode, final Class<T> nodeClassToFind) {
-        final Set<T> foundNodes = new HashSet<>();
+    public static <T extends Node, C> Set<C> findNodesOfClass(T rootNode, final Class<C> nodeClassToFind) {

```

Comment:
the idea is to have the possibility to query for something different. The was a need for this when we had a marker interface. In this case the interface was not an expression hence we could not query for it. Anyway, we can extend this a bit more by having a Predicate as second parameter? Or two version, C extends Node and Predicate.. this way could be more general.. what do you think?